openedx/core/djangoapps/course_groups/models.py

```diff
@@ -58,6 +61,111 @@ def create(cls, name, course_id, group_type=COHORT):
         )
 
 
+class CohortMembership(models.Model):
+    """Used internally to enforce our particular definition of uniqueness"""
+
+    class Meta(object):
+        unique_together = (('user', 'course_id'), )
+
+    #the sole purpose of overriding this method is to get the django 1.6 behavior of allowing 'validate_unique'
+    #for django 1.8 upgrade, just remove this method and allow the base method to be called instead.
+    #reference: https://docs.djangoproject.com/en/1.6/ref/models/instances/, under "Validating Objects"
+    def full_clean(self, **kwargs):
+        self.clean_fields()
+        self.clean()
+        if 'validate_unique' not in kwargs or kwargs['validate_unique'] is True:
+            self.validate_unique()
+
+    def clean_fields(self, *args, **kwargs):
+        if self.course_id is None:
+            self.course_id = self.course_user_group.course_id
+        super(CohortMembership, self).clean_fields(*args, **kwargs)
+
+    def clean(self):
+        if self.course_user_group.group_type != CourseUserGroup.COHORT:  # pylint: disable=E1101
+            raise ValidationError("CohortMembership cannot be used with CourseGroup types other than COHORT")
+        if self.course_user_group.course_id != self.course_id:
+            raise ValidationError("Non-matching course_ids provided")
+
+    def save(self, *args, **kwargs):
+        self.full_clean(validate_unique=False)
```

Comment:
If we don't, then an error gets thrown because we have the new value (that we're trying to save) and the old value (currently in the database) in memory at the same time. However, I needed to have the other validation code in place to ensure `CohortMembership` is only ever used with `CourseUserGroup` objects of the appropriate type.